File: GeoBroker-Server/src/test/java/de/hasenburg/geobroker/server/scenarios/PublishSubscribeTest.java
package de.hasenburg.geobroker.server.scenarios;

import de.hasenburg.geobroker.client.communication.InternalClientMessage;
import de.hasenburg.geobroker.client.main.SimpleClient;
import de.hasenburg.geobroker.commons.Utility;
import de.hasenburg.geobroker.commons.communication.ZMQProcessManager;
import de.hasenburg.geobroker.commons.model.message.ControlPacketType;
import de.hasenburg.geobroker.commons.model.message.ReasonCode;
import de.hasenburg.geobroker.commons.model.message.Topic;
import de.hasenburg.geobroker.commons.model.message.payloads.CONNECTPayload;
import de.hasenburg.geobroker.commons.model.message.payloads.PUBLISHPayload;
import de.hasenburg.geobroker.commons.model.message.payloads.SUBSCRIBEPayload;
import de.hasenburg.geobroker.commons.model.spatial.Geofence;
import de.hasenburg.geobroker.commons.model.spatial.Location;
import de.hasenburg.geobroker.server.main.Configuration;
import de.hasenburg.geobroker.server.main.server.SingleGeoBrokerServerLogic;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

public class PublishSubscribeTest {

	private static final Logger logger = LogManager.getLogger();

	private SingleGeoBrokerServerLogic serverLogic;
	private ZMQProcessManager clientProcessManager;

	@SuppressWarnings("Duplicates")
	@Before
	public void setUp() {
		logger.info("Running test setUp");

		serverLogic = new SingleGeoBrokerServerLogic();
		serverLogic.loadConfiguration(new Configuration());
		serverLogic.initializeFields();
		serverLogic.startServer();

		clientProcessManager = new ZMQProcessManager();
	}

	@After
	public void tearDown() {
		logger.info("Running test tearDown.");
		clientProcessManager.tearDown(2000);
		serverLogic.cleanUp();
	}

	@SuppressWarnings({"ConstantConditions", "OptionalGetWithoutIsPresent"})
	@Test
	public void testSubscribeInGeofence() {
		logger.info("RUNNING testSubscribeInGeofence TEST");

		// connect, ping, and disconnect
		Location l = Location.random();
		Geofence g = Geofence.circle(l, 0.4);
		Topic t = new Topic("test");

		SimpleClient client = new SimpleClient(null, "localhost", 5559, clientProcessManager);
		client.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.CONNECT, new CONNECTPayload(l)));
		client.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.SUBSCRIBE,
				new SUBSCRIBEPayload(t, g)));
		client.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.PUBLISH,
				new PUBLISHPayload(t, g, "Content")));

		Utility.sleepNoLog(500, 0);

		// validate payloads
		InternalClientMessage message = client.receiveInternalClientMessage();
		assertEquals(ControlPacketType.CONNACK, message.getControlPacketType());

		message = client.receiveInternalClientMessage();
		assertEquals(ControlPacketType.SUBACK, message.getControlPacketType());

		message = client.receiveInternalClientMessage();
		assertEquals(ControlPacketType.PUBLISH, message.getControlPacketType());
		assertEquals("Content", message.getPayload().getPUBLISHPayload().get().getContent());
		logger.info("Received published message {}", message);

		message = client.receiveInternalClientMessage();
		assertEquals(ControlPacketType.PUBACK, message.getControlPacketType());
		assertEquals(ReasonCode.Success, message.getPayload().getPUBACKPayload().get().getReasonCode());
	}

	@SuppressWarnings("ConstantConditions")
	@Test
	public void testSubscriberNotInGeofence() {
		// subscriber
		Location l = Location.random();
		Geofence g = Geofence.circle(l, 0.4);
		Topic t = new Topic("test");

		SimpleClient clientSubscriber = new SimpleClient(null, "localhost", 5559, clientProcessManager);
		clientSubscriber.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.CONNECT,
				new CONNECTPayload(Location.random()))); // subscriber not in geofence
		clientSubscriber.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.SUBSCRIBE,
				new SUBSCRIBEPayload(t, g)));

		// publisher
		SimpleClient clientPublisher = new SimpleClient(null, "localhost", 5559, clientProcessManager);
		clientPublisher.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.CONNECT,
				new CONNECTPayload(l))); // publisher is in geofence
		clientPublisher.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.PUBLISH,
				new PUBLISHPayload(t, g, "Content")));

		Utility.sleepNoLog(500, 0);

		validateNoPublishReceived(clientSubscriber, clientPublisher);
	}

	@SuppressWarnings("ConstantConditions")
	@Test
	public void testPublisherNotInGeofence() {
		// subscriber
		Location l = Location.random();
		Geofence g = Geofence.circle(l, 0.4);
		Topic t = new Topic("test");

		SimpleClient clientSubscriber = new SimpleClient(null, "localhost", 5559, clientProcessManager);
		clientSubscriber.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.CONNECT,
				new CONNECTPayload(l))); // subscriber is in geofence
		clientSubscriber.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.SUBSCRIBE,
				new SUBSCRIBEPayload(t, g)));

		// publisher
		SimpleClient clientPublisher = new SimpleClient(null, "localhost", 5559, clientProcessManager);
		clientPublisher.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.CONNECT,
				new CONNECTPayload(Location.random()))); // publisher is not in geofence
		clientPublisher.sendInternalClientMessage(new InternalClientMessage(ControlPacketType.PUBLISH,
				new PUBLISHPayload(t, g, "Content")));

		Utility.sleepNoLog(500, 0);

		validateNoPublishReceived(clientSubscriber, clientPublisher);
	}

	@SuppressWarnings("OptionalGetWithoutIsPresent")
	private void validateNoPublishReceived(SimpleClient clientSubscriber, SimpleClient clientPublisher) {
		// check subscriber messages: must not contain "PUBLISH"
		int subscriberMessageCount = 2;
		for (int i = 0; i < subscriberMessageCount; i++) {
			assertNotEquals(ControlPacketType.PUBLISH,
					clientSubscriber.receiveInternalClientMessage().getControlPacketType()); // no publish message
		}

		// check publisher messages: should contain a PUBACK with no matching subscribers
		int publisherMessageCount = 2;
		for (int i = 0; i < publisherMessageCount; i++) {
			InternalClientMessage message = clientPublisher.receiveInternalClientMessage();
			assertNotEquals(ControlPacketType.PUBLISH, message.getControlPacketType()); // no publish message
			if (i == 1) {
				assertEquals(ControlPacketType.PUBACK, message.getControlPacketType());
				assertEquals(ReasonCode.NoMatchingSubscribers,
						message.getPayload().getPUBACKPayload().get().getReasonCode());
			}
		}
	}

}

File: GeoBroker-Server/src/test/kotlin/de/hasenburg/geobroker/server/scenarios/PublishSubscribeTest.kt
package de.hasenburg.geobroker.server.scenarios

import de.hasenburg.geobroker.client.main.SimpleClient
import de.hasenburg.geobroker.commons.*
import de.hasenburg.geobroker.commons.communication.ZMQProcessManager
import de.hasenburg.geobroker.commons.model.message.Payload.*
import de.hasenburg.geobroker.commons.model.message.ReasonCode
import de.hasenburg.geobroker.commons.model.message.Topic
import de.hasenburg.geobroker.commons.model.spatial.Geofence
import de.hasenburg.geobroker.commons.model.spatial.Location
import de.hasenburg.geobroker.server.main.Configuration
import de.hasenburg.geobroker.server.main.server.SingleGeoBrokerServerLogic
import io.prometheus.client.CollectorRegistry
import org.apache.logging.log4j.LogManager
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class PublishSubscribeTest {

    private val logger = LogManager.getLogger()
    private lateinit var serverLogic: SingleGeoBrokerServerLogic

    @Before
    fun setUp() {
        logger.info("Running test setUp")
        CollectorRegistry.defaultRegistry.clear();

        serverLogic = SingleGeoBrokerServerLogic()
        serverLogic.loadConfiguration(Configuration())
        serverLogic.initializeFields()
        serverLogic.startServer()

    }

    @After
    fun tearDown() {
        logger.info("Running test tearDown.")
        serverLogic.cleanUp()
    }

    @Test
    fun testSubscribeInGeofence() {
        logger.info("RUNNING testSubscribeInGeofence TEST")

        // connect, ping, and disconnect
        val l = Location.random()
        val g = Geofence.circle(l, 0.4)
        val t = Topic("test")

        val client = SimpleClient("localhost", 5559)
        client.send(CONNECTPayload(l))
        client.send(SUBSCRIBEPayload(t, g))
        client.send(PUBLISHPayload(t, g, "Content"))

        sleepNoLog(500, 0)

        // validate received payloads
        assertTrue(client.receiveWithTimeout(100) is CONNACKPayload)
        assertTrue(client.receiveWithTimeout(100) is SUBACKPayload)
        var payload = client.receiveWithTimeout(100)
        logger.info("Received published message {}", payload)
        if (payload is PUBLISHPayload) {
            assertEquals("Content", payload.content)
        } else {
            fail("Wrong payload, received $payload")
        }

        payload = client.receiveWithTimeout(100)
        if (payload is PUBACKPayload) {
            assertEquals(ReasonCode.Success, payload.reasonCode)
        } else {
            fail("Wrong payload, received $payload")
        }

        client.tearDownClient()
    }

    @Test
    fun testSubscriberNotInGeofence() {
        // subscriber
        val l = Location.random()
        val g = Geofence.circle(l, 0.4)
        val t = Topic("test")

        val clientSubscriber = SimpleClient("localhost", 5559)
        clientSubscriber.send(CONNECTPayload(null)) // subscriber not in geofence
        clientSubscriber.send(SUBSCRIBEPayload(t, g))

        // publisher
        val clientPublisher = SimpleClient("localhost", 5559)
        clientPublisher.send(CONNECTPayload(l)) // publisher is in geofence
        clientPublisher.send(PUBLISHPayload(t, g, "Content"))

        sleepNoLog(500, 0)

        validateNoPublishReceived(clientSubscriber, clientPublisher)

        clientSubscriber.tearDownClient()
        clientPublisher.tearDownClient()
    }

    @Test
    fun testPublisherNotInGeofence() {
        // subscriber
        val l = Location.random()
        val l2 = l.locationInDistance(3000.0, 0.0)
        val g = Geofence.circle(l, 0.4)
        val t = Topic("test")

        val clientSubscriber = SimpleClient("localhost", 5559)
        clientSubscriber.send(CONNECTPayload(l)) // subscriber is in geofence
        clientSubscriber.send(SUBSCRIBEPayload(t, g))

        // publisher
        val clientPublisher = SimpleClient( "localhost", 5559)
        clientPublisher.send(CONNECTPayload(l2)) // publisher is not in geofence
        clientPublisher.send(PUBLISHPayload(t, g, "Content"))

        sleepNoLog(500, 0)

        validateNoPublishReceived(clientSubscriber, clientPublisher)

        clientSubscriber.tearDownClient()
        clientPublisher.tearDownClient()
    }

    private fun validateNoPublishReceived(clientSubscriber: SimpleClient, clientPublisher: SimpleClient) {
        // check subscriber messages: must not contain "PUBLISH"
        val subscriberMessageCount = 2
        for (i in 0 until subscriberMessageCount) {
            assertFalse(clientSubscriber.receiveWithTimeout(100) is PUBLISHPayload)
        }

        // check publisher messages: second should be a PUBACK with no matching subscribers
        assertFalse(clientPublisher.receiveWithTimeout(100) is PUBLISHPayload)

        val payload = clientPublisher.receiveWithTimeout(100)
        if (payload is PUBACKPayload) {
            assertEquals(ReasonCode.NoMatchingSubscribers, payload.reasonCode)
        } else {
            fail("Wrong payload, received $payload")
        }

    }

}
